---
title: "Infer ASVs with DADA2"
author: "Sola Takahashi"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      fig.align = "center",
                      fig.path = "../figures/01_DADA2/") # send any figure output to this folder
```

# Goals of this file

1. Use raw fastq files and generate quality plots to assess quality of reads.
2. Filter and trim out bad sequences and bases from our sequencing files.
3. Write out fastq files with high quality sequences.
4. Evaluate the quality from out filter and trim. 

# Load libraries 
```{r load-libraries}
#install.packages("devtools")
library(devtools)

#devtools::install_github("benjjneb/dada2")
library(dada2)
```



# Load Data
```{r load-data}
# Set the raw fastq path to the raw sequencing files 
# Path to the fastq files 
raw_fastqs_path <- "data/01_DADA2/01_raw_gzipped_fastqs"
raw_fastqs_path

# What files are in this path? Intuition Check
list.files(raw_fastqs_path)

# How many files are there?
str(list.files(raw_fastqs_path))
# each spot represents one fastq file 

# because paired end data
# Forward read (R1) 515F
# Reverse read (R2) 806R
# don't have primers in sequences because they have been removed, important to know if primers were included or not
# V3-V4 the overlap would be very short (430 bp). but with just V4 the overlap is majority (2 quality scores) (250 bp), can figure out true error v false error more 


# Create a vector of forward reads
# check how many files we should get
forward_reads <- list.files(raw_fastqs_path, pattern = "R1_001.fastq.gz", full.names = TRUE)

# Intution Check
head(forward_reads)

# Create a vector of reverse reads
reverse_reads <- list.files(raw_fastqs_path, pattern = "R2_001.fastq.gz", full.names = TRUE)
head(reverse_reads)


```

# Quality Plots
```{r raw-quality-plot}
# Randomly select 2 samples from dataset to evaluate
random_samples <- sample(1:length(reverse_reads), size=2)
random_samples

# Calculate and plot the quality of these two samples
plotQualityProfile(forward_reads[random_samples])
plotQualityProfile(reverse_reads[random_samples])

```







