---
title: "Phyloseq Preprocessing"
author: "Sola Takahashi"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup}
knitr::opts_chunk$set(echo = TRUE, 
                      fig.align = "center",
                      fig.path = "../figures/02_PreProcessing/") # send any figure output to this folder
```


# Load Libraries 
```{r load-libraries}
pacman::p_load(devtools, phyloseq, tidyverse, dada2, install = FALSE)

```

# Goals: 
Here we will process the data into a phyloseq object.

- ASV Table
- Taxonomy Table
- Track reads (metadata)

Then, we will remove the following: 

1. Remove Chloroplasts
2. Remove Mitochondria
3. Remove ASVs from negative controls and also negative control samples
4. Evaluate accuracy by looking at the Mock Community
5. Remove samples without enough reads

Finally, write a data file of phyloseq output, which will be stored in 'data/02_PreProcessing/raw-preprocessed-phyloseq.RData'


# Load Data 

## ASV Table
```{r load-asv-table}
# first we will load asv table 
load("data/01_DADA2/ASV_counts.RData")

#Inspect ASV tab
head(asv_tab)[,1:5]

# Fix names
sample_names <- colnames(asv_tab)
samples_fixed <- sapply(strsplit(basename(sample_names), "_"), `[`,1) 
head(samples_fixed)

# Re-write the ASV count file to fix names 
colnames(asv_tab) <- samples_fixed
str(asv_tab)

```


## Taxonomy Table
```{r load-tax-table}
tax_df <- read.table("data/01_DADA2/ASV_taxonomy.tsv", sep = "\t", skip = 1)
head(tax_df)

# fix column names
colnames(tax_df) <- c("asv_names", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species", "ASV", "ASVseq")
head(tax_df)

# Taxonomy Table Matrix
tax_mat <-
  tax_df %>% 
  tibble::column_to_rownames(., var = "asv_names") %>% 
  as.matrix()

```


## Track Reads Data
```{r load-track-reads}
load("data/01_DADA2/track_read_counts.RData")

# take a look at the data
head(track_counts_df)
dim(track_counts_df)

# Please in the terminal, copy the metadata.csv
# Into SalinityGradient_16S/ data
# from: /workdir/in_class_data/SalinityGradient_16S/
# scp /workdir/in_class_data/SalinityGradient_16S/metadata.csv  /workdir/st974/git_repo/SalinityGradient_16S/data/

# Load in the metadata
metadata_df <- read.csv("data/metadata.csv")
head (metadata_df)
dim(metadata_df)
# have 95 samples 
head(track_counts_df)

#check column names
colnames(metadata_df)

# Merge metadata_df with track_reads_df
metadata_track_reads_df <- metadata_df %>%
  left_join(., track_counts_df, by = "names") %>%
  dplyr::select(-X)

# Intuition check
head(metadata_track_reads_df)

# Update row.names to be sample names
# Before
row.names(metadata_track_reads_df)
# Rewrite
row.names(metadata_track_reads_df) <- metadata_track_reads_df$names
# Check if it worked
row.names(metadata_track_reads_df)

# Intuition Check 
head(metadata_track_reads_df)

```


# Handoff to Phyloseq
```{r phyloseq-handoff}
# Double check it's all good
dim(asv_tab) # count table
dim(tax_mat) # taxonomy matrix
# rows match

# Intuition Check
stopifnot(row.names(asv_tab) == row.names(tax_mat))
# no output = happy

# Construct the phyloseq object
raw_physeq <- phyloseq(otu_table(asv_tab, taxa_are_rows = TRUE), 
                       sample_data(metadata_track_reads_df),
                       tax_table(tax_mat))
raw_physeq

# Save this raw phyloseq object
save(raw_physeq, file = "data/02_PreProcessing/raw_physeq.RData")
```

# Clean up the data

Remove:

1. Chloroplasts
2. Mitochondria

```{r rm-mitos-chloros} 
# Remind myself of tax table
#view(tax_mat)

# Make a new physeq without chloroplasts
noChloros_physeq <- 
  raw_physeq %>%
  # rm chloroplasts
  subset_taxa(Order != "Chloroplast" | is.na(Order))
# is.na prevents over loss

# How many taxa were chloroplasts?
num_chloro_ASVs <- ntaxa(raw_physeq) - ntaxa(noChloros_physeq)
num_chloro_ASVs

# Intuition Check
#noChloros_physeq%>%
#  tax_table() %>%
#  data.frame() %>%
#  view()

# Remove Mitochondria
noChlorosMitos_physeq <- 
  noChloros_physeq %>%
  subset_taxa(Family != "Mitochondria" | is.na(Family))

# How many mitochondrial ASVs
num_mito_ASVs <- ntaxa(noChloros_physeq) - ntaxa(noChlorosMitos_physeq)
num_mito_ASVs

# How many total ASVs were removed from Chloro and Mitos?
ntaxa(raw_physeq) - ntaxa(noChlorosMitos_physeq)
# Proportions of ASVs
ntaxa(noChlorosMitos_physeq)/ntaxa(raw_physeq)
# lost about 12% of ASVs
```

# Evaluate and remove the control samples 

Take a look at the negative controls and then make a decision about whether or not to remove the ASVs found in our controls 

1. Negative Controls
2. ASVs found within the negative controls and their distribution in the samples.
3. Evaluate the mock community



```{r neg-controls} 
# Create a vector of samples that were negative controls
control_samples <- c("WaterControl", "022um-Control", 
                     "3um-Control", "DNA-Ext-Control")

# Make a new phyloseq object of only the control samples 
control_physeq <- 
  noChlorosMitos_physeq %>%
  subset_samples(., names %in% control_samples) %>%
  # force remove ASVs with 0 counts
  prune_taxa(taxa_sums(.) > 0, .)

control_physeq

# Vector of Control ASVs
control_ASVs <-
  control_physeq %>%
  tax_table() %>%
  data.frame() %>%
  dplyr::select(ASV) %>%
  as.vector()

# Visualize raw abundance values of control ASVs
control_physeq %>%
  plot_bar(., "ASV", fill = "ASV")
# ASV_3 is really abundant 
# the third most abundant ASV is found in negative control...

############# Check the control ASVs in the samples
# Make a new phyloseq object with only the control ASVs and ALL samples 
controlASV_physeq <- 
  noChlorosMitos_physeq %>%
  subset_taxa(., ASV %in% control_ASVs$ASV)

# Inspect it 
controlASV_physeq

# Let's take a look at the abundances
controlASV_physeq %>%
  plot_bar(., "ASV", fill = "ASV", facet_grid = Sample_or_Control~.)
# ASV 3, 18, 56 -> somewhat abundant in the data

# Check taxa of ASV_3
#controlASV_physeq %>%
#  tax_table() %>%
#  data.frame() %>%
#  view()
# pseudomonas -> in a lot of skin/human samples
# don't keep ASV_3
# ASV_18 is actinobacteria -> not sus
# ASV_56 also not sus

# Would only like to keep ASV_18 and ASV_56, everything else is trash
controlASVs_toREMOVE <- dplyr::setdiff(control_ASVs$ASV, c("ASV_18", "ASV_56"))
length(controlASVs_toREMOVE)

noControlsChloroMitos_physeq <- 
  noChlorosMitos_physeq %>%
  #Actually remove control ASVs except 18 and 56
  subset_taxa(., !(ASV %in% controlASVs_toREMOVE)) %>%
  # Remove control samples
  subset_samples(., !(names %in% control_samples))

# Inspect no cotrols object
noControlsChloroMitos_physeq

# After filtering how many ASVs were removed
total_ASVs_rm <- ntaxa(raw_physeq) - ntaxa(noControlsChloroMitos_physeq)
total_ASVs_rm


```

# Evaluate Accuracy in the Mock Community

- make a mock phyloseq object
- load in the standard fasta sequences of the mock community
- test for differences between the two

```{r eval-accuracy-mock}
# Make mock phyloseq object
mock_physeq <- noControlsChloroMitos_physeq %>%
  subset_samples(., names == "MockZymoPos") %>%
  prune_taxa(taxa_sums(.) >0, .)

# Inspect it
mock_physeq

tax_table(mock_physeq)

# Write us a message summarizing this
cat("DADA2 inferred", ntaxa(mock_physeq), "ASVs present in the mock community.")
# Output: DADA2 inferred 8 ASVs present in the mock community.

# Getting the mock community file
# cp /workdir/in_class_data/SalinityGradient_16S/mock_amplicons.fasta  /workdir/st974/git_repo/SalinityGradient_16S/data/02_PreProcessing/

####### Load in the mock reference sequence from Zymo 
mock_ref <- getSequences("data/02_PreProcessing/mock_amplicons.fasta")
names(mock_ref)

### Pull out seqs from Seqnenced Mock Community 
mock_seqs <- 
  mock_physeq %>%
    tax_table() %>%
    data.frame() %>%
    dplyr::select(ASV, ASVseq)

# Inspect
head(mock_seqs)

# Which ASVs match the reference?
match_asvs <- 
  mock_seqs %>%
  rowwise() %>%
  # check each ASV if they match, TRUE, if not = FALSE
  mutate(Match = any(grepl(ASVseq, mock_ref))) %>%
  # create a vector of names that matched
  pull(ASV)

# Evaluate which ones do NOT match??
# What's their abundances?
# This helps provide error rate


cat(length(match_asvs), "ASVs were exact matches to the expected reference sequence")
# mock community has a few sequences because the ASVs in match_asvs are in the 100s

# How many sequences were in the mock?
mock_seqDepth <- 
mock_physeq %>%
  otu_table() %>%
  data.frame() %>%
  colSums()
# Output: 3878

# Who is in the sequenced mock community?
mock_physeq %>%
  tax_table() %>%
  data.frame() %>%
  dplyr::select(Genus, Species)

# Curious: what are the mock counts in the actual samples?
noControlsChloroMitos_physeq %>%
  subset_taxa(., ASV %in% match_asvs) %>%
  otu_table() %>%
  t()

```

What can you conclude about the mock community???
A: 8/8 match perfectly 
A: all 8 ASVs in the mock exactly match 8 of 9 in the reference file

## Remove mock community and it's ASVs
```{r rm-mmock-asvs}
# Make a new phyloseq object without mock and its asvs
noControlsMock_ChloroMitos_physeq <- 
noControlsChloroMitos_physeq %>%
  # remove mock community 
  subset_samples(., names != "MockZymoPos") %>%
  # remove ASVs from the mock
  subset_taxa(., !(ASV %in% match_asvs))

noControlsMock_ChloroMitos_physeq

# Intuition Check
ntaxa(noControlsChloroMitos_physeq) - ntaxa(noControlsMock_ChloroMitos_physeq)

```


# Evaluate the Sequencing Depth
```{r seq-depth}
# the current data object
noControlsMock_ChloroMitos_physeq

# What is the library size/sequencing depth for each sample ?
seqSums_df <-
  noControlsMock_ChloroMitos_physeq %>%
  otu_table() %>%
  # Sum each sample column_to_rownames
  colSums() %>%
  data.frame() %>%
  rownames_to_column(var = "names") %>%
  left_join(., metadata_track_reads_df, by = "names")
  # removes the samples that don't match in metadata_track_reads_df

# Check
dim(seqSums_df)
head(seqSums_df)

# show the depth of samples
#seqSums_df %>%
 # dplyr::select(names, TotalSeqs)%>%
  #arrange(TotalSeqs) 

# Rename second column
colnames(seqSums_df)[2] <- "TotalSeqs"

# Plot it!
seqSums_df %>%
  ggplot(aes(x=reorder(names, TotalSeqs), y = TotalSeqs, fill = station)) +
  geom_bar(stat = "identity") 

# Density plot
seqSums_df %>%
  ggplot(aes(TotalSeqs, fill = station)) +
  geom_density(alpha = 0.5)

```

# Remove samples with few reads
```{r rem-samps}
# what's the min seq depth?
min(sample_sums(noControlsMock_ChloroMitos_physeq))

# remove 20210615-MA-SCB2F 
raw_preprocessed_physeq <-
  noControlsMock_ChloroMitos_physeq %>%
  subset_samples(., names != "20210615-MA-SCB2F")

# What's the new min seq depth?
min(sample_sums(raw_preprocessed_physeq))

```

# Save Preprocessed Phyloseq Object
```{r save physeq}
save(raw_preprocessed_physeq,
     file = "data/02_PreProcessing/raw_preprocessed_physeq.RData")

```



# Session Information
```{r session-info}
# Ensure reproducibility
devtools::session_info()
```



