---
title: "Between Sample (Beta) Diversity of Microbes along a Salinity Gradient"
author: "Sola Takahashi"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      fig.path = "../figures/05_CommunityAnalysis/")
```

# Goals

1. Load in the phyloseq data with rooted tree
2. Evaluate the sequencing depth and remove sample.
3. Normalize the read counts between samples.
4. Calculate community dissimilarities. Numbers between 0 and 1, completely similar versus if tehy are 1 then they are completely dissimilar. 
    a. *Sorensen*: shared species as a binary value, (abundance unweighted)
    b. *Bray-Curtis*: shared abundant species, (abundance weighted)
    c. *Weighted UNIFRAC*: consider the abundant species and where they fall on the tree, (abundance weighed)
5. Visualize the community data with two unconstrained Ordinations.
    a. *PCoA*: Linear method, Eigen value = Calculate how much variation is explained by each axis. Choose to view axis 1, 2, 3, etc and plot them together 
    b. *NMDS*: Non linear method, Multiple dimenstions into 2 or 3 axes. Need to report a stress value that's < 0.15 (how stressed the data is when it's being smooshed)
6. Run statistics with PERMANOVA and betadispR.


# Setup

## Load Libraries
```{r load-libraries}
#install.packages("vegan")
pacman::p_load(tidyverse, devtools, phyloseq, patchwork, vegan,
               install= FALSE)

# Load Station Colors
station_colors <- c(
  "Shipping Channel" = "dodgerblue4",
  "Aransas Bay" = "dodgerblue2",
  "Copano West" = "#D9CC3C",
  "Copano East" = "#A0E0BA",
  "Mesquite Bay" = "#00ADA7")
```

## Load Data
```{r load-physeq}
# Load in rooted phylogenetic tree
load("data/03_Phylogenetic_Tree/phytree_preprocessed_physeq.RData")
midroot_physeq_rm456
unrooted_physeq_rm456
```


# Explore Read Counts 
```{r calc-seq-depth}
# Calculate the total number of reads/sample
raw_TotalSeqs_df <- 
  midroot_physeq_rm456 %>%
  # calc the sample read sums
  sample_sums() %>%
  data.frame() 
# name the column
colnames(raw_TotalSeqs_df)[1] <- "TotalSeqs"
head(raw_TotalSeqs_df)

# Make a histogram of raw reads
raw_TotalSeqs_df %>%
  ggplot(aes(x = TotalSeqs)) +
  geom_histogram(bins = 50) +
  scale_x_continuous(limits = c(0, 10000)) +
  labs(title = "Raw Sequencing Depth Distribution") +
  theme_classic()


```

## Normalize Read Counts 

# Calculate and Visuallize (PCoA) community dissimilarity

## Sorensen PCoA

## Bray-Curtis PCoA

## Weighted Unifrac PCoA

## Bray-Curtis NMDS

# Test for Stastistical Significance with PERMANOVA and betadispR



# Session Information
```{r session-info}
devtools::session_info()
```


